File: packages/adapters/src/venues/nado.ts
import type { AssetClass, FundingEvent, FundingSnapshot } from "@ai-rates/core";
import type { HttpClient } from "../http";
import { marketRef, num } from "../parse";
import type { SnapshotBatch, VenueAdapter } from "../types";

/**
 * Nado (Vertex successor on Ink).
 *
 * REQUESTS: one per cycle, `GET archive /v2/contracts?edge=false` (every perp with funding, mark,
 * index, OI and volume inline), plus `GET gateway /v1/query?type=symbols` once an hour for each
 * market's `trading_status`. Queries draw on a per-IP budget of 2,400 weight a minute or 400 every 10s
 * (https://docs.nado.xyz/developer-resources/api/rate-limits); symbols weighs 2, funding history
 * 2 + limit/100. 100ms spacing cannot approach it. `edge=false` keeps volume and OI to this chain
 * ("when turned off, it only returns metrics for the current chain"); on 2026-09-13 both settings
 * returned identical BTC figures.
 *
 * ACCEPT-ENCODING: the gateway answers 403 `{"reason": "Invalid compression headers: 'Accept-Encoding'
 * must include 'gzip', 'br' or 'deflate'"}` without it (curl, 2026-09-13); the archive did not care.
 * `http.ts` already lets an adapter pass headers, and Bun's fetch still decompresses when the header is
 * set explicitly (measured: 200, `content-encoding: gzip`, JSON body). So every request sends
 * `accept-encoding: gzip` rather than relying on the runtime's default. No change to http.ts is needed.
 *
 * FUNDING: the contracts `funding_rate` is a 24-HOUR rate, and positive means longs pay. The field doc
 * (https://docs.nado.xyz/developer-resources/api/v2/contracts) says "Current 24hr funding rate. Can
 * compute hourly funding rate dividing by 24"; the funding doc (https://docs.nado.xyz/core/funding-rates)
 * computes an 8-hour F, settles F/8 every hour on the hour, and reports "the equivalent 24-hour rate --
 * three times the 8-hour rate (3F)"; non-crypto markets settle 1/24 of the same reported daily figure.
 * So the published number is kept over `basisHours` 24 and `intervalHours` is 1. It is `predicted`:
 * the archive's `funding_rate` query calls it "the latest predicted 24hr rate", and BTC moved from
 * 0.000110094 at 22:28 to 0.000103965 at 22:33 inside one hour. Checked live on 2026-09-13:
 * - quiet markets pin to the formula floor, NEAR and kPEPE 0.0003 = 3 x 0.0001, i.e. 0.0000125/h,
 *   Hyperliquid's own floor;
 * - BTC 0.000110/24 = 0.0000046/h (4.0% APR) against Hyperliquid 0.0000116/h (10.1%), ETH
 *   0.000145/24 = 0.0000060/h against 0.0000125; read as hourly it would be 96% APR for BTC;
 * - realized hourly settlements from `funding_rate_history` were 0.0000069 (22:00) and 0.0000033
 *   (21:00) for BTC, the same order as the predicted figure divided by 24;
 * - predicted against realized across the 23:00 settlement: the last reading at 22:59:33, divided by
 *   24, was BTC 0.00000490 against a realized 0.00000495, WTI -0.000135175 against -0.000134580 and
 *   ETH 0.00000212 against 0.00000193. At 23:00:30 the predicted figure had restarted for the next
 *   hour (BTC -0.000235 per day), so early-hour readings are the noisiest.
 *
 * UNITS, checked live: `open_interest_usd` is USD and equals `open_interest` (base) x mark
 * (BTC 215.3981 x 76,805.68 = $16.54M against 16,540,431). `quote_volume` is 24h volume in the quote
 * asset USDT0 (BTC $88.9M = 1,153.8 BTC x ~$77k). `next_funding_rate_timestamp` is unix SECONDS.
 *
 * TRADABILITY: contracts lists every perp, halted or not (82 on 2026-09-13). The gateway's
 * `trading_status` decides: 71 `live` are collected; 5 `post_only` (ANSEM, ARB, CASHCAT, PONS, USELESS),
 * 5 `not_tradable` (ADA, AXS, BERA, SKR, VIRTUAL) and 1 `soft_reduce_only` (PENG) are not.
 *
 * CLASS: the API declares none (contracts, symbols, all_products and v2 assets carry no category), but
 * Nado's documentation declares one per market, in the "Feed Reference" table of its oracle page
 * (https://docs.nado.xyz/llms-full.txt, "Type: Crypto, FX, Metals, Energy, US Equity, HK Equity,
 * xStock"). That declaration is copied into `NADO_DOCUMENTED_CLASSES` as it stood on 2026-09-13. A
 * market the table does not list declares nothing and is crypto, like every other venue's undeclared
 * listing, so a newly listed equity reads crypto until the table is refreshed. Of the 71 live markets on
 * 2026-09-13: crypto 40, equity 26 (25 US Equity with SPY and QQQ among them, plus HK Equity ZHIPU),
 * fx 3 (EURUSD, GBPUSD, USDJPY), commodity 2 (XAG Metals, WTI Energy). `premium_x18` is NOT a class
 * signal: it is absent on 34 markets, including crypto WLD, MEGA, SKR and CHIP, and present on HK
 * equity ZHIPU.
 *
 * QUOTE: USDT0, the `quote_currency` of every contract and the collateral funding is paid in
 * ("impacting a user's unsettled USDT0", https://docs.nado.xyz/core/funding-rates).
 *
 * BASE: the parser reads `BTC-PERP_USDT0` as BTC and agrees with `base_currency` minus `-PERP` on all
 * 82, except that it reads `kPEPE` and `kBONK` as PEPE and BONK x1000, which is what they are ("Thousand
 * Pepe Perp" in v2 assets). WTI reaches CL through core's alias, as it does for every venue.
 */

const VENUE = "nado";
export const NADO_ARCHIVE = "https://archive.prod.nado.xyz";
export const NADO_GATEWAY = "https://gateway.prod.nado.xyz/v1";
export const NADO_HEADERS: Record<string, string> = { "accept-encoding": "gzip" };
const QUOTE = "USDT0";
const RATE_BASIS_HOURS = 24;
const INTERVAL_HOURS = 1;
const SYMBOLS_TTL_MS = 3_600_000;
const HISTORY_PAGE_SIZE = 1000;
const HISTORY_MAX_PAGES = 50;

export interface NadoContract {
  product_id: number;
  ticker_id: string;
  base_currency: string;
  quote_currency?: string;
  product_type: string;
  mark_price?: number | null;
  index_price?: number | null;
  /** Base units. */
  open_interest?: number | null;
  open_interest_usd?: number | null;
  /** 24h volume in the quote asset (USDT0). */
  quote_volume?: number | null;
  /** Predicted 24-hour rate. */
  funding_rate?: number | null;
  /** Unix seconds. */
  next_funding_rate_timestamp?: number | null;
}

export interface NadoSymbol {
  type: string;
  product_id: number;
  symbol: string;
  trading_status: string;
}

export interface NadoSymbolsResponse {
  status: string;
  data: { symbols: Record<string, NadoSymbol> };
}

export interface NadoFundingHistoryResponse {
  funding_rates: { product_id: number; timestamp: string; funding_rate_x18: string }[];
}

/**
 * Per-market Type from Nado's oracle Feed Reference (https://docs.nado.xyz/llms-full.txt), 2026-09-13.
 * Only non-crypto rows are listed; the 50 Crypto perps (XAUT among them) need no entry.
 */
export const NADO_DOCUMENTED_CLASSES: Readonly<Record<string, AssetClass>> = {
  // US Equity (26)
  "AAPL-PERP": "equity",
  "AMD-PERP": "equity",
  "AMZN-PERP": "equity",
  "AVGO-PERP": "equity",
  "BBX-PERP": "equity",
  "CRCL-PERP": "equity",
  "DELL-PERP": "equity",
  "GOOGL-PERP": "equity",
  "HIMS-PERP": "equity",
  "INTC-PERP": "equity",
  "LLY-PERP": "equity",
  "META-PERP": "equity",
  "MRVL-PERP": "equity",
  "MSFT-PERP": "equity",
  "MSTR-PERP": "equity",
  "MU-PERP": "equity",
  "NBIS-PERP": "equity",
  "NVDA-PERP": "equity",
  "ORCL-PERP": "equity",
  "PENG-PERP": "equity",
  "QQQ-PERP": "equity",
  "SKHY-PERP": "equity",
  "SNDK-PERP": "equity",
  "SPCX-PERP": "equity",
  "SPY-PERP": "equity",
  "TSLA-PERP": "equity",
  // HK Equity (1)
  "ZHIPU-PERP": "equity",
  // FX (3)
  "EURUSD-PERP": "fx",
  "GBPUSD-PERP": "fx",
  "USDJPY-PERP": "fx",
  // Metals (1) and Energy (1)
  "XAG-PERP": "commodity",
  "WTI-PERP": "commodity",
};

/** The class Nado's docs declare for a perp symbol (`AAPL-PERP`); crypto when undeclared. */
export function nadoAssetClass(symbol: string): AssetClass {
  return NADO_DOCUMENTED_CLASSES[symbol] ?? "crypto";
}

function ref(contract: Pick<NadoContract, "ticker_id" | "base_currency">) {
  return marketRef(VENUE, contract.ticker_id, {
    quote: QUOTE,
    assetClass: nadoAssetClass(contract.base_currency),
  });
}

export function parseNadoSnapshots(
  contracts: Readonly<Record<string, NadoContract>>,
  symbols: Readonly<Record<string, NadoSymbol>>,
  now: number,
): FundingSnapshot[] {
  const snapshots: FundingSnapshot[] = [];
  for (const contract of Object.values(contracts)) {
    const status = symbols[contract.base_currency];
    const rate = num(contract.funding_rate);
    if (
      contract.product_type !== "perpetual" ||
      status?.type !== "perp" ||
      status.trading_status !== "live" ||
      rate === null
    ) {
      continue;
    }
    const nextSeconds = num(contract.next_funding_rate_timestamp);
    snapshots.push({
      ...ref(contract),
      observedAt: now,
      rate,
      basisHours: RATE_BASIS_HOURS,
      intervalHours: INTERVAL_HOURS,
      nextFundingAt: nextSeconds !== null && nextSeconds > 0 ? nextSeconds * 1000 : null,
      kind: "predicted",
      markPrice: num(contract.mark_price),
      indexPrice: num(contract.index_price),
      openInterestUsd: num(contract.open_interest_usd),
      volume24hUsd: num(contract.quote_volume),
    });
  }
  return snapshots;
}

/** Realized HOURLY settlements (x18 fixed point, unix seconds) within [fromMs, toMs], oldest first. */
export function parseNadoFundingHistory(
  rows: NadoFundingHistoryResponse["funding_rates"],
  contract: Pick<NadoContract, "ticker_id" | "base_currency">,
  fromMs: number,
  toMs: number,
): FundingEvent[] {
  const base = ref(contract);
  const bySettlement = new Map<number, FundingEvent>();
  for (const row of rows) {
    const seconds = num(row.timestamp);
    const x18 = num(row.funding_rate_x18);
    if (seconds === null || x18 === null) continue;
    const settledAt = seconds * 1000;
    if (settledAt < fromMs || settledAt > toMs) continue;
    bySettlement.set(settledAt, {
      ...base,
      settledAt,
      // The docs call history "realized hourly rates ... multiply by 24 for the daily equivalent".
      rate: x18 / 1e18,
      basisHours: INTERVAL_HOURS,
      markPrice: null,
    });
  }
  return [...bySettlement.values()].sort((a, b) => a.settledAt - b.settledAt);
}

export function createNadoAdapter(): VenueAdapter {
  let symbols: Record<string, NadoSymbol> | null = null;
  let symbolsFetchedAt = 0;
  /** ticker_id -> contract, for history, which is addressed by product_id. */
  const contractsByTicker = new Map<string, NadoContract>();

  async function fetchContracts(client: HttpClient): Promise<Record<string, NadoContract>> {
    const body = await client.getJson<Record<string, NadoContract>>(
      `${NADO_ARCHIVE}/v2/contracts?edge=false`,
      NADO_HEADERS,
    );
    if (!body || typeof body !== "object" || Array.isArray(body)) {
      throw new Error(`${VENUE}: unexpected contracts response`);
    }
    for (const contract of Object.values(body)) contractsByTicker.set(contract.ticker_id, contract);
    return body;
  }

  return {
    venueId: VENUE,
    minIntervalMs: 100,

    async fetchSnapshots(client, now): Promise<SnapshotBatch> {
      if (!symbols || now - symbolsFetchedAt >= SYMBOLS_TTL_MS) {
        const body = await client.getJson<NadoSymbolsResponse>(
          `${NADO_GATEWAY}/query?type=symbols`,
          NADO_HEADERS,
        );
        if (body?.status !== "success" || typeof body.data?.symbols !== "object") {
          throw new Error(`${VENUE}: unexpected symbols response`);
        }
        symbols = body.data.symbols;
        symbolsFetchedAt = now;
      }
      const contracts = await fetchContracts(client);
      return { snapshots: parseNadoSnapshots(contracts, symbols, now), settled: [] };
    },

    async fetchFundingHistory(client, venueSymbol, fromMs, toMs) {
      if (!contractsByTicker.has(venueSymbol)) await fetchContracts(client);
      const contract = contractsByTicker.get(venueSymbol);
      if (!contract) return [];
      const rows: NadoFundingHistoryResponse["funding_rates"] = [];
      let startSeconds = Math.floor(fromMs / 1000);
      const endSeconds = Math.floor(toMs / 1000);
      for (let page = 0; page < HISTORY_MAX_PAGES && startSeconds <= endSeconds; page++) {
        const body = await client.postJson<NadoFundingHistoryResponse>(
          `${NADO_ARCHIVE}/v1`,
          {
            funding_rate_history: {
              product_id: contract.product_id,
              start_time: startSeconds,
              end_time: endSeconds,
              limit: HISTORY_PAGE_SIZE,
            },
          },
          NADO_HEADERS,
        );
        const batch = body?.funding_rates;
        if (!Array.isArray(batch)) throw new Error(`${VENUE}: unexpected funding_rate_history`);
        rows.push(...batch);
        if (batch.length < HISTORY_PAGE_SIZE) break;
        // Ascending: continue from one second past the newest row, as the docs prescribe.
        const newest = Math.max(...batch.map((r) => Number(r.timestamp)));
        if (!Number.isFinite(newest)) break;
        startSeconds = newest + 1;
      }
      return parseNadoFundingHistory(rows, contract, fromMs, toMs);
    },
  };
}

export const nadoAdapter: VenueAdapter = createNadoAdapter();

File: apps/collector-go/internal/adapters/nado/nado.go
// Package nado parses Nado's perpetuals (the Vertex successor on Ink).
//
// Ported from packages/adapters/src/venues/nado.ts and pinned to the same fixtures
// (packages/adapters/__fixtures__/nado) and the same expected values as nado.test.ts, so the Go and
// TypeScript parsers cannot drift apart while both are collecting.
//
// REQUESTS: one per cycle, `GET archive /v2/contracts?edge=false` (every perp with funding, mark,
// index, OI and volume inline), plus `GET gateway /v1/query?type=symbols` once an hour for each
// market's `trading_status`. Queries draw on a per-IP budget of 2,400 weight a minute or 400 every
// 10s (https://docs.nado.xyz/developer-resources/api/rate-limits); symbols weighs 2, funding history
// 2 + limit/100. 100ms spacing cannot approach it. `edge=false` keeps volume and OI to this chain
// ("when turned off, it only returns metrics for the current chain"); on 2026-09-13 both settings
// returned identical BTC figures.
//
// ACCEPT-ENCODING: the gateway answers 403 `{"reason": "Invalid compression headers:
// 'Accept-Encoding' must include 'gzip', 'br' or 'deflate'"}` without it (curl, 2026-09-13); the
// archive did not care. The TypeScript adapter therefore sends `accept-encoding: gzip` explicitly,
// because Bun's fetch otherwise leaves the header off. Go needs no such constant and this port
// carries none: net/http's transport adds `Accept-Encoding: gzip` to every request that does not set
// one itself and transparently decompresses the reply, which is the behaviour the TypeScript header
// was there to restore. httpclient.Client exposes no per-call headers, so an explicit copy could not
// be sent anyway without changing a shared file.
//
// FUNDING: the contracts `funding_rate` is a 24-HOUR rate, and positive means longs pay. The field
// doc (https://docs.nado.xyz/developer-resources/api/v2/contracts) says "Current 24hr funding rate.
// Can compute hourly funding rate dividing by 24"; the funding doc
// (https://docs.nado.xyz/core/funding-rates) computes an 8-hour F, settles F/8 every hour on the
// hour, and reports "the equivalent 24-hour rate -- three times the 8-hour rate (3F)"; non-crypto
// markets settle 1/24 of the same reported daily figure. So the published number is kept over
// BasisHours 24 and IntervalHours is 1. It is `predicted`: the archive's `funding_rate` query calls
// it "the latest predicted 24hr rate", and BTC moved from 0.000110094 at 22:28 to 0.000103965 at
// 22:33 inside one hour. Checked live on 2026-09-13:
//   - quiet markets pin to the formula floor, NEAR and kPEPE 0.0003 = 3 x 0.0001, i.e. 0.0000125/h,
//     Hyperliquid's own floor;
//   - BTC 0.000110/24 = 0.0000046/h (4.0% APR) against Hyperliquid 0.0000116/h (10.1%), ETH
//     0.000145/24 = 0.0000060/h against 0.0000125; read as hourly it would be 96% APR for BTC;
//   - realized hourly settlements from `funding_rate_history` were 0.0000069 (22:00) and 0.0000033
//     (21:00) for BTC, the same order as the predicted figure divided by 24;
//   - predicted against realized across the 23:00 settlement: the last reading at 22:59:33, divided
//     by 24, was BTC 0.00000490 against a realized 0.00000495, WTI -0.000135175 against -0.000134580
//     and ETH 0.00000212 against 0.00000193. At 23:00:30 the predicted figure had restarted for the
//     next hour (BTC -0.000235 per day), so early-hour readings are the noisiest.
//
// UNITS, checked live: `open_interest_usd` is USD and equals `open_interest` (base) x mark
// (BTC 215.3981 x 76,805.68 = $16.54M against 16,540,431). `quote_volume` is 24h volume in the quote
// asset USDT0 (BTC $88.9M = 1,153.8 BTC x ~$77k). `next_funding_rate_timestamp` is unix SECONDS.
//
// TRADABILITY: contracts lists every perp, halted or not (82 on 2026-09-13). The gateway's
// `trading_status` decides: 71 `live` are collected; 5 `post_only` (ANSEM, ARB, CASHCAT, PONS,
// USELESS), 5 `not_tradable` (ADA, AXS, BERA, SKR, VIRTUAL) and 1 `soft_reduce_only` (PENG) are not.
//
// CLASS: the API declares none (contracts, symbols, all_products and v2 assets carry no category),
// but Nado's documentation declares one per market, in the "Feed Reference" table of its oracle page
// (https://docs.nado.xyz/llms-full.txt, "Type: Crypto, FX, Metals, Energy, US Equity, HK Equity,
// xStock"). That declaration is copied into DocumentedClasses as it stood on 2026-09-13. A market
// the table does not list declares nothing and is crypto, like every other venue's undeclared
// listing, so a newly listed equity reads crypto until the table is refreshed. Of the 71 live
// markets on 2026-09-13: crypto 40, equity 26 (25 US Equity with SPY and QQQ among them, plus HK
// Equity ZHIPU), fx 3 (EURUSD, GBPUSD, USDJPY), commodity 2 (XAG Metals, WTI Energy). `premium_x18`
// is NOT a class signal: it is absent on 34 markets, including crypto WLD, MEGA, SKR and CHIP, and
// present on HK equity ZHIPU.
//
// QUOTE: USDT0, the `quote_currency` of every contract and the collateral funding is paid in
// ("impacting a user's unsettled USDT0", https://docs.nado.xyz/core/funding-rates).
//
// BASE: the parser reads `BTC-PERP_USDT0` as BTC and agrees with `base_currency` minus `-PERP` on
// all 82, except that it reads `kPEPE` and `kBONK` as PEPE and BONK x1000, which is what they are
// ("Thousand Pepe Perp" in v2 assets). WTI reaches CL through core's alias, as it does for every
// venue.
package nado

import (
	"bytes"
	"encoding/json"
	"errors"
	"sort"

	"github.com/belyjelli/ai-rates/collector/internal/adapters"
	"github.com/belyjelli/ai-rates/collector/internal/core"
)

const VenueID = "nado"

const (
	// quote is USDT0 on every contract, and is passed as an override because the symbol parser does
	// not recognise it: "BTC-PERP_USDT0" splits to a second token of USDT0, which is not one of the
	// quotes core knows, so the parsed quote is nil.
	quote = "USDT0"
	// rateBasisHours: `funding_rate` is the predicted TWENTY-FOUR hour rate. Read as hourly, BTC
	// would annualise at 96% instead of 4%.
	rateBasisHours = 24.0
	// intervalHours: settlement is every hour on the hour, of 1/24 of the reported daily figure.
	intervalHours = 1.0
	// x18Scale is the fixed-point divisor on the history endpoint's `funding_rate_x18`.
	x18Scale = 1e18
)

// Contract is one row of the archive's /v2/contracts object.
type Contract struct {
	ProductID     int    `json:"product_id"`
	TickerID      string `json:"ticker_id"`
	BaseCurrency  string `json:"base_currency"`
	QuoteCurrency string `json:"quote_currency"`
	ProductType   string `json:"product_type"`

	MarkPrice  adapters.Num `json:"mark_price"`
	IndexPrice adapters.Num `json:"index_price"`
	// OpenInterest is BASE units; OpenInterestUSD is the same figure already priced.
	OpenInterest    adapters.Num `json:"open_interest"`
	OpenInterestUSD adapters.Num `json:"open_interest_usd"`
	// QuoteVolume is 24h volume in the quote asset (USDT0).
	QuoteVolume adapters.Num `json:"quote_volume"`
	// FundingRate is the PREDICTED 24-hour rate.
	FundingRate adapters.Num `json:"funding_rate"`
	// NextFundingRateTimestamp is unix SECONDS, not milliseconds.
	NextFundingRateTimestamp adapters.Num `json:"next_funding_rate_timestamp"`
}

// Contracts is the archive's /v2/contracts object flattened to a slice, in the order the response
// lists it.
//
// The archive keys contracts by ticker and the TypeScript side reads them with Object.values(),
// which is document order. A Go map would hand back a different order on every cycle for no gain --
// the key is never read, since every row repeats it as `ticker_id` -- so the decode keeps the
// document's order and drops the keys.
type Contracts []Contract

// Symbol is one entry of the gateway's `symbols` map. `trading_status` is the only thing that says
// whether a listed perp is actually tradable.
type Symbol struct {
	Type          string `json:"type"`
	ProductID     int    `json:"product_id"`
	Symbol        string `json:"symbol"`
	TradingStatus string `json:"trading_status"`
}

// SymbolsResponse is the gateway's /query?type=symbols body. Symbols is a pointer so an absent or
// null `symbols` is distinguishable from an empty one, which is what the TypeScript
// `typeof body.data?.symbols !== "object"` guard turns into a thrown error.
type SymbolsResponse struct {
	Status string `json:"status"`
	Data   struct {
		Symbols *map[string]Symbol `json:"symbols"`
	} `json:"data"`
}

// FundingHistoryRow is one realized hourly settlement. Both numerics arrive as quoted strings:
// `timestamp` in unix SECONDS and `funding_rate_x18` in 1e18 fixed point.
type FundingHistoryRow struct {
	ProductID      int          `json:"product_id"`
	Timestamp      adapters.Num `json:"timestamp"`
	FundingRateX18 adapters.Num `json:"funding_rate_x18"`
}

// FundingHistoryResponse is the archive's `funding_rate_history` reply. The slice is a pointer
// because a missing array is a broken response, not an empty page.
type FundingHistoryResponse struct {
	FundingRates *[]FundingHistoryRow `json:"funding_rates"`
}

// ErrUnexpectedContracts, ErrUnexpectedSymbols and ErrUnexpectedHistory mirror the TypeScript
// adapter's throws. A response that lost its payload has to fail the venue's cycle rather than read
// as a venue with nothing listed and no funding ever settled.
var (
	ErrUnexpectedContracts = errors.New("nado: unexpected contracts response")
	ErrUnexpectedSymbols   = errors.New("nado: unexpected symbols response")
	ErrUnexpectedHistory   = errors.New("nado: unexpected funding_rate_history")
)

// UnmarshalJSON walks the contracts object as tokens, keeping the document's order. A JSON array
// where an object belongs is the TypeScript `Array.isArray(body)` rejection.
func (c *Contracts) UnmarshalJSON(data []byte) error {
	decoder := json.NewDecoder(bytes.NewReader(data))
	token, err := decoder.Token()
	if err != nil {
		return err
	}
	if delim, ok := token.(json.Delim); !ok || delim != '{' {
		return ErrUnexpectedContracts
	}

	contracts := make(Contracts, 0, 96)
	for decoder.More() {
		// The key is the ticker, which the row repeats as `ticker_id`; read past it.
		if _, err := decoder.Token(); err != nil {
			return err
		}
		var contract Contract
		if err := decoder.Decode(&contract); err != nil {
			return err
		}
		contracts = append(contracts, contract)
	}
	*c = contracts
	return nil
}

// DocumentedClasses is the per-market Type from Nado's oracle Feed Reference
// (https://docs.nado.xyz/llms-full.txt), 2026-09-13, keyed by `base_currency` (the perp symbol,
// "AAPL-PERP"). Only non-crypto rows are listed; the 50 Crypto perps (XAUT among them) need no
// entry.
var DocumentedClasses = map[string]core.AssetClass{
	// US Equity (26)
	"AAPL-PERP":  core.ClassEquity,
	"AMD-PERP":   core.ClassEquity,
	"AMZN-PERP":  core.ClassEquity,
	"AVGO-PERP":  core.ClassEquity,
	"BBX-PERP":   core.ClassEquity,
	"CRCL-PERP":  core.ClassEquity,
	"DELL-PERP":  core.ClassEquity,
	"GOOGL-PERP": core.ClassEquity,
	"HIMS-PERP":  core.ClassEquity,
	"INTC-PERP":  core.ClassEquity,
	"LLY-PERP":   core.ClassEquity,
	"META-PERP":  core.ClassEquity,
	"MRVL-PERP":  core.ClassEquity,
	"MSFT-PERP":  core.ClassEquity,
	"MSTR-PERP":  core.ClassEquity,
	"MU-PERP":    core.ClassEquity,
	"NBIS-PERP":  core.ClassEquity,
	"NVDA-PERP":  core.ClassEquity,
	"ORCL-PERP":  core.ClassEquity,
	"PENG-PERP":  core.ClassEquity,
	"QQQ-PERP":   core.ClassEquity,
	"SKHY-PERP":  core.ClassEquity,
	"SNDK-PERP":  core.ClassEquity,
	"SPCX-PERP":  core.ClassEquity,
	"SPY-PERP":   core.ClassEquity,
	"TSLA-PERP":  core.ClassEquity,
	// HK Equity (1)
	"ZHIPU-PERP": core.ClassEquity,
	// FX (3)
	"EURUSD-PERP": core.ClassFX,
	"GBPUSD-PERP": core.ClassFX,
	"USDJPY-PERP": core.ClassFX,
	// Metals (1) and Energy (1)
	"XAG-PERP": core.ClassCommodity,
	"WTI-PERP": core.ClassCommodity,
}

// AssetClassFor is the class Nado's docs declare for a perp symbol ("AAPL-PERP"); crypto when
// undeclared, like every other venue's undeclared listing.
func AssetClassFor(symbol string) core.AssetClass {
	if class, declared := DocumentedClasses[symbol]; declared {
		return class
	}
	return core.ClassCrypto
}

// refFor builds a market reference from the ticker, with the quote and the documented class the
// venue's own data does not put on the symbol.
func refFor(contract Contract) core.MarketRef {
	class := AssetClassFor(contract.BaseCurrency)
	q := quote
	return adapters.MarketRefFor(VenueID, contract.TickerID, adapters.Overrides{
		Quote:      &q,
		HasQuote:   true,
		AssetClass: &class,
	})
}

// ParseSnapshots normalises one cycle's contracts against the gateway's trading statuses.
//
// A market the gateway does not list, or lists as anything but a `live` perp, is skipped: contracts
// carries halted markets too, and the status is the only signal that separates them.
func ParseSnapshots(contracts Contracts, symbols map[string]Symbol, now int64) []core.FundingSnapshot {
	snapshots := make([]core.FundingSnapshot, 0, len(contracts))
	for _, contract := range contracts {
		status, listed := symbols[contract.BaseCurrency]
		rate := contract.FundingRate
		if contract.ProductType != "perpetual" ||
			!listed || status.Type != "perp" || status.TradingStatus != "live" ||
			!rate.OK {
			continue
		}

		// SECONDS, so a straight copy would file every settlement in 1970.
		var nextFundingAt *int64
		if next := contract.NextFundingRateTimestamp; next.OK && next.Val > 0 {
			at := int64(next.Val) * 1000
			nextFundingAt = &at
		}

		interval := intervalHours
		snapshots = append(snapshots, core.FundingSnapshot{
			MarketRef:  refFor(contract),
			ObservedAt: now,
			Rate:       rate.Val,
			// The published figure is a daily rate settled in 24 hourly slices, so the basis is 24
			// and the interval is 1. Collapsing the two would annualise BTC at 96% instead of 4%.
			BasisHours:      rateBasisHours,
			IntervalHours:   &interval,
			NextFundingAt:   nextFundingAt,
			Kind:            core.KindPredicted,
			MarkPrice:       contract.MarkPrice.Ptr(),
			IndexPrice:      contract.IndexPrice.Ptr(),
			OpenInterestUSD: contract.OpenInterestUSD.Ptr(),
			Volume24hUSD:    contract.QuoteVolume.Ptr(),
		})
	}
	return snapshots
}

// ParseFundingHistory returns realized HOURLY settlements (x18 fixed point, unix seconds) within
// [fromMs, toMs], oldest first.
func ParseFundingHistory(rows []FundingHistoryRow, contract Contract, fromMs, toMs int64) []core.FundingEvent {
	base := refFor(contract)

	// Keyed by settlement so a row repeated across a page boundary lands once, the later read
	// winning, exactly as the TypeScript Map does.
	bySettlement := make(map[int64]core.FundingEvent, len(rows))
	for _, row := range rows {
		if !row.Timestamp.OK || !row.FundingRateX18.OK {
			continue
		}
		settledAt := int64(row.Timestamp.Val) * 1000
		if settledAt < fromMs || settledAt > toMs {
			continue
		}
		bySettlement[settledAt] = core.FundingEvent{
			MarketRef: base,
			SettledAt: settledAt,
			// The docs call history "realized hourly rates ... multiply by 24 for the daily
			// equivalent", so unlike the snapshot's daily figure this one's basis is the hour.
			Rate:       row.FundingRateX18.Val / x18Scale,
			BasisHours: intervalHours,
			MarkPrice:  nil,
		}
	}

	settlements := make([]int64, 0, len(bySettlement))
	for settledAt := range bySettlement {
		settlements = append(settlements, settledAt)
	}
	sort.Slice(settlements, func(i, j int) bool { return settlements[i] < settlements[j] })

	events := make([]core.FundingEvent, 0, len(settlements))
	for _, settledAt := range settlements {
		events = append(events, bySettlement[settledAt])
	}
	return events
}
